js/compiler.ts

```diff
@@ -1,76 +1,43 @@
 import * as ts from "typescript";
-import * as msg from "gen/cli/msg_generated";
-import { window } from "./window";
-import { assetSourceCode } from "./assets";
+import * as os from "./os";
 import { bold, cyan, yellow } from "./colors";
-import { Console } from "./console";
+import { window } from "./window";
+import { postMessage, workerClose, workerMain } from "./workers";
 import { core } from "./core";
+import { Console } from "./console";
+import { assert, notImplemented } from "./util";
 import { cwd } from "./dir";
-import { sendSync } from "./dispatch";
-import * as flatbuffers from "./flatbuffers";
-import * as os from "./os";
-import { TextDecoder, TextEncoder } from "./text_encoding";
-import { clearTimer, setTimeout } from "./timers";
-import { postMessage, workerClose, workerMain } from "./workers";
-import { assert, log, notImplemented } from "./util";
-
-const EOL = "\n";
-const ASSETS = "$asset$";
-const LIB_RUNTIME = `${ASSETS}/lib.deno_runtime.d.ts`;
+import { assetSourceCode } from "./assets";
 
-// An instance of console
+// Startup boilerplate. This is necessary because the compiler has its own
+// snapshot. (It would be great if we could remove these things or centralize
+// them somewhere else.)
 const console = new Console(core.print);
+window.console = console;
+window.workerMain = workerMain;
+export default function denoMain(): void {
+  os.start("TS");
+}
 
-/** The location that a module is being loaded from. This could be a directory,
- * like `.`, or it could be a module specifier like
- * `http://gist.github.com/somefile.ts`
- */
-type ContainingFile = string;
-/** The internal local filename of a compiled module. It will often be something
- * like `/home/ry/.deno/gen/f7b4605dfbc4d3bb356e98fda6ceb1481e4a8df5.js`
- */
-type ModuleFileName = string;
-/** The original resolved resource name.
- * Path to cached module file or URL from which dependency was retrieved
- */
-type ModuleId = string;
-/** The external name of a module - could be a URL or could be a relative path.
- * Examples `http://gist.github.com/somefile.ts` or `./somefile.ts`
- */
-type ModuleSpecifier = string;
-/** The compiled source code which is cached in `.deno/gen/` */
-type OutputCode = string;
-/** The original source code */
-type SourceCode = string;
-/** The output source map */
-type SourceMap = string;
+const ASSETS = "$asset$";
+const OUT_DIR = "$deno";
 
 /** The format of the work message payload coming from the privileged side */
-interface CompilerLookup {
-  specifier: ModuleSpecifier;
-  referrer: ContainingFile;
-  cmdId: number;
-}
-
-/** Abstraction of the APIs required from the `os` module so they can be
- * easily mocked.
- */
-interface Os {
-  fetchModuleMetaData: typeof os.fetchModuleMetaData;
-  exit: typeof os.exit;
-  noColor: typeof os.noColor;
+interface CompilerReq {
+  rootNames: string[];
+  // TODO(ry) add compiler config to this interface.
```

Comment:
I think that is going to be hard, if not impossible to do in Rust.  Mostly because TypeScript supports JSONC (JSON with comments) and uses its own JSON parsers to accomplish this.  serde does not support JSONC, so I don't think there is any solution but to pass a string/u8 buffer in the request.  Technically there are some other things that the parser would resolve (like extending tsconfigs, project references, etc) which we don't have implemented which doesn't really make it practical to parse config on the Rust side as well.

Reply:
> JSON parsers to accomplish this. serde does not support JSONC. serde does not support JSONC

Ok good to know. I guess we'll cross that bridge when we get there.